### PR TITLE
[feat] Add Support Cut-Cross-Entropy (CCE) 

### PR DESCRIPTION
**Pull request by w1ida**
This PR adds optional **Cut-Cross-Entropy (CCE)** loss into the ms-swift training pipeline.
CCE is a memory-efficient alternative to standard CE, from Apple’s project:

➡️ [https://github.com/apple/ml-cross-entropy](https://github.com/apple/ml-cross-entropy)

---

## ✨ Features

* Add `use_cce=True` in `TrainArguments` to enable CCE.
* Plug-and-play replacement; no model modification required.
* Add test script: `tests/train/test_cce.py`.

**Usage:**

```python
TrainArguments(
    model='Qwen/Qwen2.5-0.5B-Instruct',
    dataset=['gsm8k#1024'],
    split_dataset_ratio=0.01,
    use_cce=True,
    **kwargs
)
```

Install dependency:

```bash
pip install "cut-cross-entropy @ git+https://github.com/w1ida/ml-cross-entropy.git"
```

---

## 📊 Memory Result (A10, BS=64)

* **Hardware:** NVIDIA A10
* **transformers:** 4.56.1
* **Model:** `Qwen/Qwen2.5-0.5B-Instruct`
* **Dataset:** `gsm8k#1024`
* **Batch size:** 64

### **Environment**

* **Hardware:** NVIDIA A10
* **transformers:** 4.56.1
* **Model:** `Qwen/Qwen2.5-0.5B-Instruct`
* **Dataset:** `gsm8k#1024`
* **Batch size:** 64

### **Memory Usage**

| Loss            | Memory (GiB) |
| --------------- | ------------ |
| transformers CE | 16.24        |
| **CCE**         | **2.62**     |

**≈ 84% memory reduction.**

---

## 📄 Reference

**Cut Your Losses in Large-Vocabulary Language Models**
[https://arxiv.org/abs/2411.09009](https://arxiv.org/abs/2411.09009)

image

### CODE REVIEW

**Review by tastelikefeet — APPROVED** [review]

**Comment by Jintao-Huang** [comment]
哈咯，这是我这测试的情况，看起来liger_kernel的显存占用更少

使用的是你给的例子

"""
liger_kernel:
{'loss': 0.72876334, 'grad_norm': 5.18441868, 'learning_rate': 9.989e-05, 'epoch': 0.06, 'global_step/max_steps': '1/48', 'percentage': '2.08%', 'elapsed_time': '3s', 'remaining_time': '2m 27s', 'memory(GiB)': 6.26, 'train_speed(iter/s)': 0.318979}
{'loss': 0.55229092, 'grad_norm': 1.04675198, 'learning_rate': 9.735e-05, 'epoch': 0.31, 'global_step/max_steps': '5/48', 'percentage': '10.42%', 'elapsed_time': '8s', 'remaining_time': '1m 9s', 'memory(GiB)': 6.27, 'train_speed(iter/s)': 0.622059}
cce:
{'loss': 0.72674572, 'grad_norm': 5.16368008, 'learning_rate': 9.989e-05, 'epoch': 0.06, 'global_step/max_steps': '1/48', 'percentage': '2.08%', 'elapsed_time': '5s', 'remaining_time': '4m 3s', 'memory(GiB)': 6.74, 'train_speed(iter/s)': 0.193078}
{'loss': 0.55218643, 'grad_norm': 1.03230286, 'learning_rate': 9.735e-05, 'epoch': 0.31, 'global_step/max_steps': '5/48', 'percentage': '10.42%', 'elapsed_time': '11s', 'remaining_time': '1m 39s', 'memory(GiB)': 6.74, 'train_speed(iter/s)': 0.430464}
原始
{'loss': 0.72737986, 'grad_norm': 5.16535616, 'learning_rate': 9.989e-05, 'token_acc': 0.83662044, 'epoch': 0.06, 'global_step/max_steps': '1/48', 'percentage': '2.08%', 'elapsed_time': '2s', 'remaining_time': '1m 44s', 'memory(GiB)': 61.58, 'train_speed(iter/s)': 0.450271}
{'loss': 0.55272961, 'grad_norm': 1.02546477, 'learning_rate': 9.735e-05, 'token_acc': 0.85563783, 'epoch': 0.31, 'global_step/max_steps': '5/48', 'percentage': '10.42%', 'elapsed_time': '7s', 'remaining_time': '1m 0s', 'memory(GiB)': 61.58, 'train_speed(iter/s)': 0.714248}
"""

**Comment by w1ida** [comment]
> 哈咯，这是我这测试的情况，看起来liger_kernel的显存占用更少
> 
> 使用的是你给的例子
> 
> """ liger_kernel: {'loss': 0.72876334, 'grad_norm': 5.18441868, 'learning_rate': 9.989e-05, 'epoch': 0.06, 'global_step/max_steps': '1/48', 'percentage': '2.08%', 'elapsed_time': '3s', 'remaining_time': '2m 27s', 'memory(GiB)': 6.26, 'train_speed(iter/s)': 0.318979} {'loss': 0.55229092, 'grad_norm': 1.04675198, 'learning_rate': 9.735e-05, 'epoch': 0.31, 'global_step/max_steps': '5/48', 'percentage': '10.42%', 'elapsed_time': '8s', 'remaining_time': '1m 9s', 'memory(GiB)': 6.27, 'train_speed(iter/s)': 0.622059} cce: {'loss': 0.72674572, 'grad_norm': 5.16368008, 'learning_rate': 9.989e-05, 'epoch': 0.06, 'global_step/max_steps': '1/48', 'percentage': '2.08%', 'elapsed_time': '5s', 'remaining_time': '4m 3s', 'memory(GiB)': 6.74, 'train_speed(iter/s)': 0.193078} {'loss': 0.55218643, 'grad_norm': 1.03230286, 'learning_rate': 9.735e-05, 'epoch': 0.31, 'global_step/max_steps': '5/48', 'percentage': '10.42%', 'elapsed_time': '11s', 'remaining_time': '1m 39s', 'memory(GiB)': 6.74, 'train_speed(iter/s)': 0.430464} 原始 {'loss': 0.72737986, 'grad_norm': 5.16535616, 'learning_rate': 9.989e-05, 'token_acc': 0.83662044, 'epoch': 0.06, 'global_step/max_steps': '1/48', 'percentage': '2.08%', 'elapsed_time': '2s', 'remaining_time': '1m 44s', 'memory(GiB)': 61.58, 'train_speed(iter/s)': 0.450271} {'loss': 0.55272961, 'grad_norm': 1.02546477, 'learning_rate': 9.735e-05, 'token_acc': 0.85563783, 'epoch': 0.31, 'global_step/max_steps': '5/48', 'percentage': '10.42%', 'elapsed_time': '7s', 'remaining_time': '1m 0s', 'memory(GiB)': 61.58, 'train_speed(iter/s)': 0.714248} """

@Jintao-Huang  嗯嗯，liger_kernel 这边还额外优化了 RMSNorm、RoPE、SwiGLU 等算子，从整体显存占用来看确实更低。 [paper](https://arxiv.org/abs/2411.09009)里只对 logits + CE 部分的显存开销做了对比。
另外，[axolotl](https://github.com/axolotl-ai-cloud/axolotl) 这个框架已经支持了 CCE 和 liger-kernel，感觉可以考虑在 ms-swift 里也集成一下CCE的支持~

**Comment by Jintao-Huang** [comment]
支持CCE会额外引入维护成本，我和其他开发者讨论一下
